src/transaction/build.rs

```diff
@@ -7,228 +7,680 @@
 type MaybeCallFuture<T, R> = MaybeReady<CompatCallFuture<T, R>>;
 
+/// Type alias for an optional call future.
+type OptionCallFuture<T, R> = OptionFuture<CompatCallFuture<T, R>>;
+
+/// Type alias for future retrieving default local account parameters.
+type LocalParamsFuture<T> = Join<MaybeCallFuture<T, Vec<Address>>, OptionCallFuture<T, U256>>;
+
+/// A future for building a locally signed transaction.
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+#[pin_project]
+pub struct LocalBuildFuture<T: Transport> {
+    /// The partial transaction object with the optional condition.
+    tx: Option<(PartialTransaction, Option<TransactionCondition>)>,
+    /// The inner future for retrieving the list of accounts on the node and
+    /// gas price estimation.
+    #[pin]
+    params: LocalParamsFuture<T>,
+}
+
+impl<T: Transport> LocalBuildFuture<T> {
+    /// Create a new future for building a locally singed transaction request
+    /// from a partial transaction object and account information.
+    pub fn new(
+        web3: &Web3<T>,
+        tx: PartialTransaction,
+        from: Option<Address>,
+        condition: Option<TransactionCondition>,
+    ) -> Self {
+        let eth = web3.eth();
+        let accounts = maybe!(from.map(|from| vec![from]), eth.accounts().compat());
+        let gas_price_estimate = match tx.gas_price {
+            Some(GasPrice::Factor(_)) => Some(eth.gas_price().compat()),
+            _ => None,
+        };
+
+        LocalBuildFuture {
+            tx: Some((tx, condition)),
+            params: future::join(accounts, gas_price_estimate.into()),
+        }
+    }
+}
+
+impl<T: Transport> Future for LocalBuildFuture<T> {
+    type Output = Result<TransactionRequest, ExecutionError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut this = self.project();
+        this.params
+            .as_mut()
+            .poll(cx)
+            .map(|(accounts, gas_price_estimate)| {
+                let (tx, condition) = this.tx.take().expect("future polled more than once");
+
+                let from = accounts?.get(0).copied().unwrap_or_default();
+                let gas_price_estimate = gas_price_estimate.transpose()?;
+
+                Ok(tx.into_request(from, condition, gas_price_estimate))
+            })
+    }
+}
+
+/// A future for building a locally signed transaction with a locked account.
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+#[pin_project]
+pub struct LockedBuildFuture<T: Transport> {
+    /// The underlying web3 provider.
+    web3: Web3<T>,
+    /// The password used for signing.
+    password: Protected,
+    /// The state of the build future.
+    #[pin]
+    state: LockedBuildState<T>,
+}
+
+/// The state of the `LockedBuildFuture`.
+#[pin_project]
+enum LockedBuildState<T: Transport> {
+    /// Preparing the transaction request.
+    Preparing(#[pin] LocalBuildFuture<T>),
```

Comment:
Should be `EstimatingGas` as we only care about the gas estimate here when building transactinos this way.